iodata/formats/mwfn.py

```diff
@@ -0,0 +1,367 @@
+    #  then uses `backspace`, then begins reading. Despite this flexibility, it is stated by
+    #  the authors that the order of section, and indeed, entries in general, must be fixed.
+    #  With this in mind the input utilized some hardcoding since order should be fixed.
+    #
+    #  mwfn ignores lines beginning with `#`.
+    # read sections of mwfn file
+    # This assumes title is on first line which seems to be the standard
+    title = next(lit).strip()
+    opener_keywords = ["Wfntype", "Charge", "Naelec", "Nbelec", "E_tot", "VT_ratio", "Ncenter"]
+    wfntype, charge, nelec_a, nelec_b, \
+        energy, vt_ratio, num_atoms = _load_helper_opener(lit, opener_keywords)
+    # coordinates are in Angstrom in MWFN
+    atnums, atcorenums, atcoords = _load_helper_atoms(lit, num_atoms)
+    nbasis, nindbasis, nprim, nshell, nprimshell = _load_helper_basis(lit)
+    keywords = ["Shell types", "Shell centers", "Shell contraction"]
+    shell_types, shell_centers, prim_per_shell = _load_helper_shells(lit, nshell, keywords)
+    # HORTON indices start at 0 because Pythons do.
+    shell_centers -= 1
+    assert wfntype < 5
+    assert num_atoms > 0
+    assert min(atnums) >= 0
+    assert len(shell_types) == nshell
+    assert len(shell_centers) == nshell
+    assert len(prim_per_shell) == nshell
+    exponent = _load_helper_prims(lit, nprimshell)
+    coeffs = _load_helper_prims(lit, nprimshell)
+    # number of MO's should equal number of independent basis functions. MWFN inc. virtual orbitals.
+    num_coeffs = nindbasis
+    if wfntype in [0, 2, 3]:
+        # restricted wave function
+        num_mo = nindbasis
+    elif wfntype in [1, 4]:
+        # unrestricted wavefunction
+        num_mo = 2 * nindbasis
+
+    mo_numbers = np.empty(num_mo, int)
+    mo_type = np.empty(num_mo, int)
+    mo_occs = np.empty(num_mo, float)
+    mo_sym = np.empty(num_mo, str)
+    mo_energies = np.empty(num_mo, float)
+    mo_coeffs = np.empty([num_coeffs, num_mo], float)
+
+    for mo in range(num_mo):
+        mo_numbers[mo], mo_occs[mo], mo_energies[mo], mo_coeffs[:, mo], \
+            mo_type[mo], mo_sym[mo] = _load_helper_mo(lit, num_coeffs)
+
+    # TODO add density matrix and overlap
+
+    return {'title': title, 'energy': energy, 'wfntype': wfntype,
+            'nelec_a': nelec_a, 'nelec_b': nelec_b, 'charge': charge,
+            'atnums': atnums, 'atcoords': atcoords, 'atcorenums': atcorenums,
+            'nbasis': nbasis, 'nindbasis': nindbasis, 'nprims': nprim,
+            'nshells': nshell, 'nprimshells': nprimshell, 'full_virial_ratio': vt_ratio,
+            'shell_centers': shell_centers, 'shell_types': shell_types,
+            'prim_per_shell': prim_per_shell, 'exponents': exponent, 'coeffs': coeffs,
+            'mo_numbers': mo_numbers, 'mo_occs': mo_occs, 'mo_energies': mo_energies,
+            'mo_coeffs': mo_coeffs, 'mo_type': mo_type, 'mo_sym': mo_sym}
+
+
+def build_obasis(shell_map: np.ndarray, shell_types: np.ndarray,
```

Comment:
unless you think this might be useful for other formats:
```suggestion
def _build_obasis(shell_map: np.ndarray, shell_types: np.ndarray,
```

Reply:
The use of the build_obasis was an effort to follow the wfn.py format as a template. The actual code inside it is based on fchk.py, found on line 131 of fchk.py, modified for the differences found in mwfn. I doubt the build_obasis will be generally useful, and will rename to _build_obasis.